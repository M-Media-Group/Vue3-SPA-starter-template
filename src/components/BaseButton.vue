<script setup lang="ts">
defineProps({
  to: {
    type: String,
    required: false,
  },
});
</script>
<template>
  <a role="button" v-if="$attrs.href">
    <slot />
  </a>

  <router-link v-else-if="to" :to="to" custom v-slot="{ navigate }">
    <button v-bind="$attrs" @click="navigate">
      <slot />
    </button>
  </router-link>

  <button v-else>
    <slot />
  </button>
</template>
