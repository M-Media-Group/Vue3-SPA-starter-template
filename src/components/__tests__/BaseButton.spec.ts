import { describe, expect, it } from "vitest";

import { RouterLinkStub, mount } from "@vue/test-utils";
import BaseButton from "../BaseButton.vue";

describe("Base Button", () => {
  it("renders as a link when an href attribute is set", () => {
    const wrapper = mount(BaseButton, {
      props: {
        href: "https://google.com",
      },

      slots: {
        default: "Click me",
      },
    });

    // Expect an anchor element to exist
    expect(wrapper.find("a").exists()).toBe(true);

    // Expect the anchor element to have the correct href
    expect(wrapper.find("a").attributes("href")).toBe("https://google.com");

    // Expect the anchor element to have the correct text
    expect(wrapper.find("a").text()).toBe("Click me");
  });

  it("renders as a router-link when to attribute is set", () => {
    const wrapper = mount(BaseButton, {
      props: {
        to: "/",
      },

      slots: {
        default: "Click me",
      },
    });

    // Expect a button element to exist
    expect(wrapper.findComponent(RouterLinkStub).props().to).toBe("/");

    // Expect the element to have the correct text
    expect(wrapper.findComponent(RouterLinkStub).text()).toBe("Click me");
  });

  it("renders as a button when no href or to attribute is set", () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: "Click me",
      },
    });

    // Expect a button element to exist
    expect(wrapper.find("button").exists()).toBe(true);

    // Expect the element to have the correct text
    expect(wrapper.find("button").text()).toBe("Click me");
  });

  it("renders as a button and passes type attribute", () => {
    const wrapper = mount(BaseButton, {
      props: {
        type: "submit",
      },

      slots: {
        default: "Click me",
      },
    });

    // Expect a button element to exist
    expect(wrapper.find("button").exists()).toBe(true);

    // Expect the element to have the correct text
    expect(wrapper.find("button").attributes("type")).toBe("submit");
  });
});
