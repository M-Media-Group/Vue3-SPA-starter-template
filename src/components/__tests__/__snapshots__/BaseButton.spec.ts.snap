// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base Button > renders as a button and passes type attribute 1`] = `"<button type="submit">Click me</button>"`;

exports[`Base Button > renders as a button when no href or to attribute is set 1`] = `"<button type="button">Click me</button>"`;

exports[`Base Button > renders as a link when an href attribute is set 1`] = `
"<a role="button" href="https://google.com">
  <!-- @slot This is the text or content that is clickable and navigates. This is the slot inside the \`<a>\` tag if href is present. This component automatically sets the href or to attributes for navigation for you. It renders as an \`<a>\` tag with a \`role=button\` if its \`href\`, otherwise its a button if \`to\` is passed. -->Click me
</a>"
`;

exports[`Base Button > renders as a router-link when to attribute is set 1`] = `
"<button type="button">
  <!-- @slot This is the default slot which contains the text.  -->Click me
</button>"
`;
